### PR TITLE
feat(docs-infra): generate Angular CLI command reference

### PR DESCRIPTION
**Pull request by petebacondarwin**
This is a first cut - for demoing and discussion - of the rendering of the Angular CLI command reference in the angular.io docs website.

@Brocco  - I couldn't add you as an assignee. But I think you should be involved 🤗

### CODE REVIEW

File: aio/tools/transforms/cli-docs-package/clone-cli-repo.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const shelljs = require('shelljs');
+const {resolve}  = require('canonical-path');
+
+const REPO = 'https://github.com/hansl/angular-cli';
+const BRANCH_OR_TAG = 'commands-json-schema';
```
Comment (petebacondarwin) [inline aio/tools/transforms/cli-docs-package/clone-cli-repo.js]: Note that `REPO` is currently pointing to @hansl's fork because the JSON files have not yet landed in the main repo. Eventually this will change to `https://github.com/angular/angular-cli`.

Also, `BRANCH_OR_TAG` is currently pointing to the branch for the PR containing the JSON files, but eventually this will need to be set to the tagged commit that we want to generate CLI docs for. E.g. a release tag.

File: aio/tools/transforms/cli-docs-package/index.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  convertToJsonProcessor.docTypes = convertToJsonProcessor.docTypes.concat(['cli-command', 'cli-overview']);
+  postProcessHtml.docTypes = postProcessHtml.docTypes.concat(['cli-command', 'cli-overview']);
+});;
+
+function getCommandPaths(basePath) {
+  const COMMANDS_PATH = resolve(basePath, 'commands.json');
+  const COMMANDS_FILE = readFileSync(COMMANDS_PATH, 'utf8');
+  const commands = json.parse(COMMANDS_FILE);
+  return Object.keys(commands).map(name => resolve(basePath, commands[name]));
+}
```
Comment (petebacondarwin) [inline aio/tools/transforms/cli-docs-package/index.js]: We read the paths to each of the command JSON files from the `commands.json` file that must be in a reliable place.

File: aio/tools/transforms/cli-docs-package/processors/filterHiddenCommands.spec.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  it('should run after the correct processor', () => {
+    const processor = processorFactory();
+    expect(processor.$runAfter).toBeUndefined();
+  });
+
+  it('should run before the correct processor', () => {
+    const processor = processorFactory();
+    expect(processor.$runBefore).toEqual(['processCliContainerDoc']);
+  });
+});
```
Comment (petebacondarwin) [inline aio/tools/transforms/cli-docs-package/processors/filterHiddenCommands.spec.js]: TODO: add actual tests of this processors functionality.

File: aio/tools/transforms/cli-docs-package/processors/processCliContainerDoc.spec.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  it('should run after the correct processor', () => {
+    const processor = processorFactory();
+    expect(processor.$runAfter).toEqual(['extra-docs-added']);
+  });
+
+  it('should run before the correct processor', () => {
+    const processor = processorFactory();
+    expect(processor.$runBefore).toEqual(['rendering-docs']);
+  });
+});
```
Comment (petebacondarwin) [inline aio/tools/transforms/cli-docs-package/processors/processCliContainerDoc.spec.js]: TODO: add actual tests of this processors functionality.

File: aio/tools/transforms/cli-docs-package/readers/cli-command.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        arguments.push(propertyInfo);
+      } else {
+        options.push(propertyInfo);
+      }
+    }
+  });
+  // Ensure the argument array is in the correct order
+  arguments.sort((a,b) => a.$default.index - b.$default.index);
+  return { arguments, options };
+}
```
Comment (petebacondarwin) [inline aio/tools/transforms/cli-docs-package/readers/cli-command.js]: We may need to augment this processing to capture additional information about the constraints on the properties.

File: aio/tools/transforms/cli-docs-package/.gitignore [hunk 6]
```diff
@@ -0,0 +1 @@
+tmp
```
Comment (petebacondarwin) [inline aio/tools/transforms/cli-docs-package/.gitignore]: We clone the Angular CLI repo into this `tmp` folder, which will be ignored by git.

File: aio/tools/transforms/templates/cli/cli.template.html [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  <tr>
+    <td><a class="code-anchor" href="{$ command.path $}">{$ command.name $}</a></td>
+    <td>{% for alias in command.commandAliases %}<code>{$ alias $}</code>{% endfor %}</td>
+    <td>
+      {$ command.shortDescription | marked $}
+    </td>
+  </tr>
+{% endfor %}
+</tbody>
+</table>
```
Comment (petebacondarwin) [inline aio/tools/transforms/templates/cli/cli.template.html]: The name of this template matches the `id` of the cli overview document. This has a higher priority in the dgeni template finder than the docType template (content.template.html) which is normally used for docs such as guides.

**Comment by JoostK** [comment]
This is very cool!

@petebacondarwin I am missing a lot of CLI options for each task, e.g. [`serve` only shows `configuration` and `prod`](https://pr25363-ca3b449.ngbuilds.io/cli/serve). Is that something that could be improved?

**Comment by hansl** [comment]
@petebacondarwin Would it be useful to add hints for default schematics and default build system? So that you can add those as a "if you didn't change your angular.json, it will run the default schematics which has the following options"?

Maybe useful as a talking point in our next meeting.

**Comment by alexeagle** [comment]
Update from meeting yesterday: the static files checked into the CLI repo aren't sufficient to generate help - you need to dynamically run the schematics code to stitch in the options.
We'll run `ng help` as part of the CI build for the CLI, and publish a JSON-formatted output for all the help commands in our package snapshots.
Then the dgeni process for angular.io can fetch our snapshot repo and will have the complete data there.

**Comment by hansl** [comment]
TOOL-166 and TOOL-167 have been created to track the work from the CLI part (JSON help, and snapshots).

File: aio/content/cli.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  https://github.com/angular/angular-cli/tree/master/packages/angular/cli
+
+For details about changes between versions, and information about updating from previous releases, see the Releases tab on GitHub.
+
+### Basic workflow
+
+Invoke the tool on the command line through the ng executable. Online help is available on the command line:
+
+```
+> ng help		  	        Lists commands with short descriptions
```
Comment (gkalpak) [inline aio/content/cli.md]: Nit: This is not aligned in the resulting HTML.

Comment (petebacondarwin) [reply aio/content/cli.md]: Cool. This will be a rewrite when we get the JSON help files from the CLI.

**Comment by jenniferfell** [comment]
Quickly making sure to capture the nav things that we discussed when this first was available. 

Under CLI in the navigation, it would be nice to have children:

CLI command reference
- Introduction (in-page title "Introduction to the CLI")
- add
- build
- config
- etc.

**Comment by jenniferfell** [comment]
@petebacondarwin We have a draft of the desired output, including suboptions/commands of schematics commands. It's been in review since June. I'll repost the link in our slack channel.

**Comment by petebacondarwin** [comment]
OK, so check this out. I have updated the code to pull in the JSON help from the Angular CLI.
The current rendered output is not yet right but you can get the basic idea in the preview: https://pr25363-6f4544b.ngbuilds.io/cli/add.

**Comment by petebacondarwin** [comment]
TODO:

* remove alias columns
* add positional stuff to schematics
* flow the string options on the same line
* remove `true|false` and replace with `--flag` or `--no-flag` depending upon default.
* add a syntax line to schematics

**Comment by jenniferfell** [comment]
Quick thought: Perhaps we should alphabetize the schematics in commands such as ng generate?

File: aio/tools/transforms/cli-docs-package/processors/processCliCommands.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  };
+};
+
+
+function processOptions(container, options) {
+  container.positionalOptions = [];
+  container.namedOptions = [];
+  container.options = options.filter(option => !option.hidden);
+  container.options.forEach(option => {
+    if (!option.hidden) {
```
Comment (hansl) [inline aio/tools/transforms/cli-docs-package/processors/processCliCommands.js]: This is redundant with the above `filter(...)` on line 39.

File: aio/tools/transforms/cli-docs-package/processors/processCliContainerDoc.spec.js (hunk 4 above)
Comment (petebacondarwin) [reply aio/tools/transforms/cli-docs-package/processors/processCliContainerDoc.spec.js]: Done

File: aio/content/cli/tmp/package.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "@angular/cli": "https://github.com/angular/cli-builds#master"
```
Comment (petebacondarwin) [inline aio/content/cli/tmp/package.json]: The trick is that we should change the branch here to match the branch that we are releasing angular.io from.

**Comment by jenniferfell** [comment]
In current preview, syntax for aliases is not showing up. Instead full syntax shows up twice. 

Also from Judy in slack: 
Neither syntax statement shows the argument, which I would expect to be the arg name in italic: `ng generate *schematic* [*options*]`
...or, since there are both command options as subcommand options, perhaps `ng generate *schematic* [*command_options*][*subcommand_options*]`
The italics distinguish placeholders (as opposed to constant values like option names), so I think they are especially important the formal syntax statement. Other commands do have named arguments that would show up here - `name=*app_name*` (edited)
I think it works well to have the sub-command options on the same page.

**Comment by jenniferfell** [comment]
Can we alphabetize the schematics in commands like ng generate?

**Comment by petebacondarwin** [comment]
@jenniferfell - the subcommands are now alphabetised. I have also fixed the bugs that @jbogarthyde noted. Also I made a few changes in line with your interview with @johnpapa. Notably I have implemented all of the rendering options for options [sic]. You can choose which we through away.

**Review by jenniferfell — APPROVED** [review]
Nav looks good! Command reference pages look good for RC. We can continue our group discussions/refinements during RC period. Thanks!!

File: aio/package.json [hunk 11]
```diff
@@ -41,7 +42,8 @@
     "postcheck-env": "yarn aio-check-local",
     "payload-size": "scripts/payload.sh",
     "predocs": "yarn generate-stackblitz && yarn generate-zips",
-    "docs": "dgeni ./tools/transforms/angular.io-package",
+    "docs": "yarn docs-only",
+    "docs-only": "dgeni ./tools/transforms/angular.io-package",
```
Comment (gkalpak) [inline aio/package.json]: Why this change?

File: aio/tools/transforms/cli-docs-package/extract-cli-commands.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+const shelljs = require('shelljs');
+const {resolve}  = require('canonical-path');
+const {CONTENTS_PATH} = require('../config');
+
+shelljs.cd(resolve(CONTENTS_PATH, 'cli-src'));
+shelljs.exec('git clean -Xfd');
+shelljs.exec('yarn install');
```
Comment (gkalpak) [inline aio/tools/transforms/cli-docs-package/extract-cli-commands.js]: This script doesn't seem to extract anything 🤔
Consider renaming to something more descriptive.

**Review by gkalpak — APPROVED** [review]
LGTM

For the record, I will say that I am not a big fun of adding the `CLI COMMANDS` as a top-level menu in the sidenav. I would rather have a `CLI` submenu (above `API`).

File: aio/package.json (hunk 11 above)
Comment (petebacondarwin) [reply aio/package.json]: Often I didn't want the overhead of the Stackblitz and Zip file creation just to iterate on my changes to the doc-gen.

File: aio/tools/transforms/cli-docs-package/extract-cli-commands.js (hunk 12 above)
Comment (petebacondarwin) [reply aio/tools/transforms/cli-docs-package/extract-cli-commands.js]: It extracts the CLI library from npm :-P
Suggestion for a better name?

**Comment by petebacondarwin** [comment]
I believe that the navigation side of things is still under discussion.

**Comment by petebacondarwin** [comment]
We will need to create a new PR for patch, since the `aio/content/cli-src/package.json` needs changing to map to the correct version of Angular CLI for those docs.

**Comment by petebacondarwin** [comment]
OK, so this is only for master, which is why we can experiment with the nav a bit.
